숫자게임 [STEP 1] sookoong, Groot

Groot-94 opened the pull request:
@dnwhd0112 
안녕하세요 우롱차 Step0, 1에 대한 PR 보내드립니다 ㅎㅎ 리뷰 부탁해요 😀

# [ 고민한 부분 ]

## 1. Set 인덱스 여부 오인

순서도를 작성할 때 Set을 이용해서 숫자의 값과 위치를 비교하려고 했었습니다. 순서도 작성 시엔 문제가 없다고 생각했는데, 직접 코딩을 하면서 생각해보니 Set은 인덱스가 의미가 없어서 자리를 비교할 수 없었습니다. 그래서 Array를 Set으로 변경하려고 보니 변경 시에 정렬이 되는 문제가 생겨서, 결국 Set을 Array로 먼저 변환하고 숫자의 자리를 비교하는 방법을 사용했습니다. 순서도를 그리면서 중복을 제거해준다는 Set 컬렉션의 기능을 생각했지만, 순서가 없다는 특징을 잊고 그려서 이런 문제점이 있었던 것 같습니다.

## 2. While 이중 조건 오류

숫자야구 STEP1 중 게임 종료의 경우는 3스트라이크가 나오거나 주어진 9번의 기회에 정답을 찾지 못하는 경우가 있었습니다. 그런데 while strike != 3 || count != 0 으로 코드를 입력하였을 때, count가 0이 되었을 때는 정상적으로 loop에서 빠져나올 수 있었던 반면, 스트라이크가 3개가 나와도 사용자 승리로 숫자 야구가 종료되지 않았습니다. 이에 groot와 고민한 결과 while의 조건은 해당 조건이 아닌 경우 반복문을 실행하는 것이므로 합집합이 아닌 교집합으로 조건을 묶었을 때 최소 하나의 조건을 만족할 경우 loop가 끝날 것라는 생각을 도출하였습니다. 다시 말해, Strike가 3이거나 count가 0일 때 게임을 멈추기 위해서는 strike가 3이 아니면서 count가 0이 아닌 경우를 동시에 만족할 때 반복문이 실행되게 하면 됨을 의미합니다. 이에 따라 while strike != 3 && count != 0로 코드를 수정하니 각각의 조건에서도 잘 적용되었습니다.

--- 

# [ 궁금한 부분 ]

## 1. 문장 간 띄어쓰기 guideline

함수 내 다양한 기능들을 구현하다보니 필연적으로 코드가 많이 작성되었습니다. 이에 각각의 기능별로 이를 띄우고 사용해야할 지, 혹은 모두 붙여서 사용하는 것이 더욱 좋은 코드 작성이 될 수 있을지 고민이 많이 되었습니다. 저와 groot는 이에 language guide에서 붙여서 쓰는 것을 보고 후자의 의견이 맞다고 생각하여 Step1을 구현했습니다. Swift guidline에 따르면 문장 간 띄어쓰는 가이드라인에 대해서는 설명이 부족한 것 같아 이 판단이 맞는지 확신이 들지 않기에  이렇게 질문드립니다.

## 2. 코드 들여쓰기

숫자야구 가이드 라인에 따르면 ‘코드 들여쓰기를 2번을 초과하지 않습니다’라고 명시되었습니다. 이에 저희는 29번째 줄의 compareNumber() 함수 중 35번째 줄에 작성한 for문이 들여쓰기를 초과한 것이 아닌지 질문드립니다. for문이 만약 들여쓰기 횟수를 초과한 것이라면, 저희는 각각의 경우를 하나하나 작성해야 하는지 혹은 더 좋은 방법이 있을지 궁금하여 질문드립니다. 


(기존)
```swift 
        for number in 0...2 {
            if randomArray[number] == inputArray[number] {
                strike += 1
            }
        }
```
(변경)
```swift
     if randomArray[0] == inputArray[0] {
                strike += 1
        }
    if randomArray[1] == inputArray[1] {
                strike += 1
        }
    if randomArray[2] == inputArray[2] {
                strike += 1
        }
```
감사합니다 !

Jeon-Minsu committed:
Flow Chart.drawio

Jeon-Minsu committed:
ios-number-baseball-flowchart.drawio

Jeon-Minsu committed:
Update Untitled Diagram.drawio

Jeon-Minsu committed:
ios-number-baseball-flowchart.drawio.png

Groot-94 committed:
feat: 임의의 정수를 담을 전역변수 추가

Jeon-Minsu committed:
feat: 남은 시도횟수를 담아둘 count 변수 생성

Groot-94 committed:
refactor: RandomSet의 선언을 상수에서 변수로 변경

Groot-94 committed:
feat: 3개의 임의의 정수를 생성하여 중복을 제거하는 Set 타입으로 반환하는 함수 생성

Jeon-Minsu committed:
feat: 전역변수 strike, ball 초기화

Jeon-Minsu committed:
feat: 1~9 사이의 세 개의 임의의 정수를 생성하여 반환하는 함수생성

Groot-94 committed:
refactor: 변수타입 변경과 변수명 수정

- randomSet의 타입을 Set에서 Array로 수정
- randomSet의 이름을 randomArray 변경
- inputArray의 타입을 Set에서 Array로 수정

Jeon-Minsu committed:
refactor: 반환타입 수정

- 기존의 insertRandomNumber()을 Array(insertRandomNumber())로 수정
- 기존의 Set(unwrappedUserNumber)을 unwrappedUserNumber로 수정

Groot-94 committed:
feat: 랜덤으로 생성된 배열의 숫자와 사용자에게 입력받은 숫자를 비교해 게임을 진행하는 기능추가

Jeon-Minsu committed:
feat: 숫자야구 게임을 실행하는 시작 함수 생성

- randomArray 배열에 랜덤한 숫자를 생성하는 배열을 설정한다.
- 컴퓨터가 설정한 임의의 숫자와 사용자가 입력하는 숫자의 값을 비교하는 compareNumber함수를 입력한다.

Groot-94 committed:
refactor: 함수안에 있는 공백제거, 전역변수 userNumber 제거, 변수위치 조정, compareNumber함수 안에 있는 numberBaseball() 제거

- numberBaseball 함수는 사용자의 입력을 받기 때문에 제출을 위해 제거함.

Jeon-Minsu committed:
refactor: 함수 이름 변경 및 사용자 입력을 대신하여 임의의 변수 생성

- 기존의 numberBaseball에서 inputUserNumber로 함수명을 수정하였다.
- Step1에서 사용자 입력 대신 임의의 수로 생성하기를 요구하여 inputArray 값에 Array(insertRandomNumber())을 사용하였다.

Groot-94 committed:
docs: Step1의 순서도 README에 업로드

Groot-94 committed:
docs: Step2을 위한 순서도 수정

Jeon-Minsu committed:
docs: Step2 순서도 유효성 검사 로직 구축

Groot-94 committed:
docs: Step2 순서도 정리 및 유효성 검사 로직 수정

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,53 @@
+}
+
+func startGame() {
+    randomArray = Array(insertRandomNumber())
```
set을 받아서 Array로 변환해주는 이유가 있나요? 형 변환 없이 insertRandomNumber가 쓰이는곳이 있을까요?
그리고 Array에서 사용할수 있는 함수들도 있습니다! 중복을 체크하기 위해서라면 Array의 함수를 찾아보는것도 좋아요.

Jeon-Minsu replied:
generateRandomNumber를set으로 출력한 이후 모두 Array로 변환하여 사용하여 generateRandomNumber 함수에서 미리 Array로 return 될 수 있도록 수정하였습니다!

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,53 @@
+}
+
+func inputUserNumber() {
+    let userNumber = readLine()?.split(separator: " ").map {Int($0) ?? 0}
```
고차함수를 쓰셨군요! 아주 좋습니다!

Jeon-Minsu replied:
감사합니다!! 앞으로 더욱 더 좋은 코드를 써보도록 노력하겠습니다.

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,53 @@
 
-print("Hello, World!")
 
+func insertRandomNumber() -> Set<Int> {
```
좀 더 좋은 네이밍이 있을까요? 해당 메소드는 랜덤 넘버를 삽입해주는것 처럼 읽혀서요.

Jeon-Minsu replied:
기존의 insertRandomNumber 함수명을 generateRandomNumber로 수정하여 직관성을 높여보았습니다!

Jeon-Minsu replied:
이외에도 가독성을 위해 기존의 count를 remainingChances로, strike와 ball 변수를 strikeCount와 ballCount로 수정 완료하였습니다.

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,53 @@
+}
+
+func compareNumber() {
+    while strike != 3 && count != 0 {
```
부등호를 쓰는 방법도 좋을꺼같아요. 읽는것도 편하고요.

Jeon-Minsu replied:
compareNumber 함수 내 strikeCount와 remainingChances의 조건을 '!='을 사용하여 가독성이 떨어진 다는 점을 보완하기 위하여 부등호를 사용해 strikeCount < 3 && remainingChances > 0 으로 수정하였습니다.

dnwhd0112 commented on README.md:
```diff
@@ -4,3 +4,8 @@
+# Step0
+## Step1 순서도
+
+![ios-number-baseball-flowchart](https://user-images.githubusercontent.com/96932116/163082482-cd2b4b5a-3b4b-4bce-883a-db694f9e85fa.png)
```
일부분은 글로 풀어썻는대 어느부분은 코드를 쓰셨군요! 하나의 방법으로 통일해주시는게 읽기 편할것 같아요!
내용은 훌륭합니다.

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,53 @@
+                strike += 1
+            }
+        }
+        ball = Set(randomArray).intersection(inputArray).count - strike
```
이부분은 읽기가 좀 힘드네요 왠만하면 한 줄에 . 한개만 사용하는쪽이 읽기가 편해요.
아님 변수를 새로 만들어서 가독성을 높이는 쪽도 고려하면 좋습니다.

Jeon-Minsu replied:
기존의 ballCount에 intersection을 이용하여 한줄에 너무 긴 코드를 작성하여 이를 해결하고자 intersectionOfArray란 상수를 만들어 코드를 분리하여 다시 작성하였습니다!

dnwhd0112 reviewed (changes requested):
안녕하세요 그루트 수꿍 수고하셨습니다!
이번 프로젝트도 잘 부탁드려요.

---
```문장 간 띄어쓰기 guideline```
저는 가이드라인의 목적은 가독성을 증가시키기 위해서 라고 생각합니다. 띄어쓰는 편이 훨씬 가독성이 좋다면 띄어쓰는게 맞다고 생각해요. 저도 띄어쓰는 편입니다.

```코드 들여쓰기```
저는 기존의 방법이 더 좋다고 생각해요. 하지만 for문 안쪽에서 if문을 쓰지않는 다른 방법은 없을까요? 한번 고민해보시고 알려주세요!

저는 조금 더 기능들을 함수로 나누면 좀 더 좋은 코드가 될꺼라고 생각합니다. 한번 고민해보세요!

Jeon-Minsu committed:
refactor: Step1 Pull Request를 반영한 코드 수정

- 기존의 insertRandomNumber 함수명을 generateRandomNumber로 수정하여 직관성을 높임
- generateRandomNumber를set으로 출력한 이후 모두 Array로 변환하여 사용하여 generateRandomNumber 함수에서 미리 Array로 return 될 수 있도록 수정
- 가독성을 위해 기존의 count를 remainingChances로, strike와 ball 변수를 strikeCount와 ballCount로 수정
- compareNumber 함수 내 strikeCount와 remainingChances의 조건을 '!='을 사용하여 가독성이 떨어져 strikeCount < 3 && remainingChances > 0 으로 수정
- 기존의 ballCount에 intersection을 이용하여 한줄에 너무 긴 코드를 작성하여 이를 해결하고자 intersectionOfArray란 상수를 만들어 코드를 분
- 55번과 56번의 print함수의 중복을 피하고자 seperator을 이용하여 하나의 코드로 수정
- 가독성을 위해 코드간 간격을 생성

는

Groot-94 committed:
refactor: 스트라이크, 볼의 개수를 판단하는 기능 분리

- strikeCount() 함수 추가 및 스트라이크 판단방법을 수정
- ballCount() 함수 추가
